cloud_functions/climateiq_data_export_cf/main.py

```diff
@@ -211,19 +228,173 @@ def _build_spatialized_model_predictions(
+    spatialized_predictions = spatialized_predictions[
+        spatialized_predictions.apply(
+            lambda cell: chunk_boundary.contains(
+                geometry.Point(cell["h3_centroid_lon"], cell["h3_centroid_lat"])
+            ),
+            axis=1,
+        )
+    ]
+
+    # Extract rows where the projected H3 cell is not fully contained within the chunk
+    # so we can aggregate prediction values across chunk boundaries.
+    boundary_h3_cells = (
+        spatialized_predictions[
+            spatialized_predictions.apply(
+                lambda row: not row["h3_boundary"].within(chunk_boundary),
+                axis=1,
+            )
+        ]
+        .groupby(["h3_index"])
+        .prediction.agg("mean")
+    )
+
+    # TODO: Add logic for fetching predictions from neighboring chunks for boundary
+    # cells.
+    spatialized_predictions = spatialized_predictions.groupby(
+        ["h3_index"]
+    ).prediction.agg("mean")
+    return spatialized_predictions
+
+
+def _get_neighboring_chunks(
+    study_area_metadata: dict, chunk_metadata: dict, boundary_h3_cells: pd.DataFrame
+):
+    x = chunk_metadata["x_index"]
+    y = chunk_metadata["y_index"]
+
+    neighbors = {(x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1)}
+    queue = queue.Queue()
+    for neighbor in neighbors:
+        queue.put(neighbor)
+
+    while not queue.empty():
+        neighbor_x, neighbor_y = queue.get()
+        query = (
+            chunks_ref.where("x_index", "==", neighbor_x)
+            .where("y_index", "==", neighbor_y)
+            .limit(1)
+        )
+        chunk_doc = query.get()
+        if not chunk_doc.exists:
+            # Some chunks may not have neighbors in the study area.
+            continue
+
+        neighbor_chunk_metadata = chunk_doc.to_dict()
+        if (
+            neighbor_chunk_metadata.get("id") is None
+            or neighbor_chunk_metadata.get("row_count") is None
+            or neighbor_chunk_metadata.get("col_count") is None
+            or neighbor_chunk_metadata.get("x_ll_corner") is None
+            or neighbor_chunk_metadata.get("y_ll_corner") is None
+            or neighbor_chunk_metadata["x_index"] is None
+            or neighbor_chunk_metadata["y_index"] is None
+        ):
+            raise ValueError(
+                f'Neighbor chunk "{neighbor_chunk_metadata["id"]}" is missing one or more required '
+                "fields: id, row_count, col_count, x_ll_corner, y_ll_corner, x_index, y_index"
+            )
+
+        neighbor_chunk_boundary = _get_chunk_boundary(
+            study_area_metadata, neighbor_chunk_metadata
+        )
+        intersects = False
+        for row in boundary_h3_cells.iterrows():
+            if row["h3_boundary"].intersects(neighbor_chunk_boundary):
+                intersects = True
+                break  # Once an overlap is found, no need to check further
+
+        if intersects:
+            queue.put((neighbor_x - 1, neighbor_y))
```

Comment:
It looks like this would process every chunk in a single cloud function invocation?

My suggestion is to:
- only load direct neighbors of the chunk being processed
- only attempt to find h3 cells at the seams between the chunk being processed and its neighbors
- only look at h3 cells whose center is in the chunk being processed

this way you're still processing the chunks in parallel, only loading neighbors for those cases where you need to read over the seams in chunks, and not producing duplicate h3 cells because the center can only be in one chunk.

Reply:
hmm neighboring chunks would only be loaded if intersects = True, wdym by every chunk would get processed?

the algorithm here is:

1. process current chunk + project ML predicted cells to H3 cells
2. from the projected H3 cells, keep only the H3 cells with centroids in the current chunk
3. from the H3 cells with centroids in the current chunk, extract the boundary H3 cells (cells that intersect with neighboring chunks)
4. look at direct neighbors of current chunk:
    4a) if the current neighbor intersects with any of the boundary H3 cells, process the current neighbor's neighbors

Reply:
okay, sorry, I think I misread this. 
So we would only extend to a neighbor-of-a-neighbor if an H3 cell from the original file is large enough that it extends over multiple neighbors?

Reply:
Okay, sorry, I misread this. I thought we were doing this recursively for each neighbor, but I see we're only working with boundary_h3_cells from the original file.

Reply:
so rn I extend to a neighbor-of-a-neighbor if the current neighbor had an intersection with an H3 cell from the chunk being processed by the cloud function (without confirming if the H3 cell would also overlap with the neighbor-of-a-neighbor), do you think its worth optimizing this?

Reply:
I wouldn't spend too much time optimizing this on the first go. I'd just focus on correctness & completeness first and move onto optimization later.